src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresAccessPolicyDAO.scala

```diff
@@ -744,23 +744,53 @@ class PostgresAccessPolicyDAO(protected val writeDbRef: DbReference, protected v
   override def deleteAllResourcePolicies(resourceId: FullyQualifiedResourceId, samRequestContext: SamRequestContext): IO[Unit] = {
     val p = PolicyTable.syntax("p")
     val g = GroupTable.syntax("g")
+    val gm = GroupMemberTable.syntax("gm")
 
-    serializableWriteTransaction("deletePolicy", samRequestContext)({ implicit session =>
-      // first get all the group PKs associated to all the resource policies
-      // then delete all the resource policies
-      // then delete all the groups from the first query
-
-      val fromAndWhereFragment = samsqls"""from ${PolicyTable as p} where ${p.resourceId} = (${loadResourcePKSubQuery(resourceId)})"""
-
-      val groupPKsToDelete = samsql"""select ${p.result.groupId} $fromAndWhereFragment""".map(_.get[GroupPK](p.resultName.groupId)).list().apply()
-
-      samsql"""delete $fromAndWhereFragment""".update().apply()
+    val fromAndWhereFragment = samsqls"""from ${PolicyTable as p} where ${p.resourceId} = (${loadResourcePKSubQuery(resourceId)})"""
 
-      if (groupPKsToDelete.nonEmpty) {
-        samsql"""delete from ${GroupTable as g}
+    for {
+      groupPKsToDelete <- readOnlyTransaction("deletePolicyFindGroups", samRequestContext) { implicit session =>
+          samsql"""select ${p.result.groupId} $fromAndWhereFragment""".map(_.get[GroupPK](p.resultName.groupId)).list().apply()
+        }
```

Comment:
I know the intention here was to split this into multiple transactions, but I'm not sure that's a good idea either.  Technically, if someone did something like adding a Policy to this resource between these transactions, then we would fail to properly clean up the Group behind that new Policy.  

If we had a cascading delete on the `sam_resource_policy.group_id` then I think we would be OK and we could just delete the policies and not have to worry about also deleting the groups, but that does not seem to be the case.  I do not know why the cascading delete is missing, so we would need to do some research before considering adding it.  

It seems like we should be able to read the `groupPKsToDelete` from within a single write transaction, but I don't know much about Slick and don't know what you were wrestling with, so I appreciate that it might not seem as straight forward as I think.

Reply:
you could just get `groupPKstoDelete` again in the error handling transaction. Not the most efficient, but at this point things have already gone wrong so as long as it doesn't take down Sam it's not like users will mind if an error gets back to them slightly slower.

Agreed that it'd be nice to have the cascading delete, but like Greg I can't remember why we omitted it in the first place and I'd want to do a bit of research to be sure we aren't overlooking some important reason

Reply:
I know it's a pain, but I think it's worth keeping these queries in the same transaction, and following Marc's suggestion of rerunning the query when generating the error message.  Not ideal, but should work.

As for trying to get the intermediate result of `groupPKstoDelete`, I would try starting the transaction and then try using `zip` per https://scala-slick.org/doc/3.2.0/dbio.html#sequential-execution.  I'm not 100% that that will work for you, but that's where I'd start at least.

Reply:
totally makes sense. I combined the transactions and pulled out the extracted the query into a val and now it is run twice.